### PR TITLE
CI: use ocaml 4.14.0 for edge switch

### PR DESCRIPTION
**Pull request by SkySkimmer**
Overlays:
- https://github.com/LPCIC/coq-elpi/pull/366
- https://github.com/maximedenes/vscoq/pull/8
- https://gitlab.inria.fr/gappa/coq/-/merge_requests/9

### CODE REVIEW

**Comment by SkySkimmer** [comment]
What should we do about the Stream deprecation?
- put disabled deprecation attributes where it's used
- depend on camlp-streams (can we do that for just 4.14?)
- vendor
- other?

**Comment by Alizter** [comment]
Is there any downside to depending on camp-streams for every version?

**Comment by ejgallego** [comment]
@ppedrot @herbelin ?

**Comment by ejgallego** [comment]
Personally I think that's a good candidate for vendoring, as we may want to have full control of this type, and likely at some point implement coq-specific functionality.

**Comment by Alizter** [comment]
@ejgallego What's wrong with putting it in clib?

**Comment by gares** [comment]
Why do we use streams in the first place? Is it a leftover from camlp5?

**Comment by ppedrot** [comment]
It seems to be only used by lexing / parsing indeed. I'd recommend vendoring, it's a small module.

**Comment by gares** [comment]
If possible, vendor it in gramlib, so that it is clear it is not to be used for other purposes

**Comment by SkySkimmer** [comment]
Let's see how this goes

File: tools/coqdep/common.ml [hunk 1]
```diff
@@ -248,6 +248,7 @@ let parse_META package f =
     close_in ic;
     Some (f, m)
   with
+  (* with findlib >= 1.9.3 this should be Fl_metascanner.Error *)
   | Stream.Error msg -> Error.cannot_parse_meta_file package msg
```
Comment (SkySkimmer) [inline tools/coqdep/common.ml:252]: Our current minimum bound for findlib is 1.9.1 which doesn't have the Fl_metascanner.Error exception.
This means that when using 1.9.3 we don't catch findlib parsing errors.
Should we change our minimum findlib version?

Comment (gares) [reply tools/coqdep/common.ml:252]: ok

**Comment by ejgallego** [comment]
> @ejgallego What's wrong with putting it in clib?

I don't want `clib` to grow, but to get smaller as we have better things to use the time than maintaining our own stdlib.

**Comment by ejgallego** [comment]
> Putting it in gramlib doesn't work well, it's also used by coqproject_file which is in lib/ (and used by coqdep, coqide, coq_makefile)

Put in in `vendor/cstream`

**Comment by SkySkimmer** [comment]
Putting it in clib works fine and I don't want to debug some other setup.
I also don't see how adding a new directory makes anything easier.

File: tools/coqdep/common.ml (hunk 1 above)
Comment (SkySkimmer) [reply tools/coqdep/common.ml]: bumped

**Review by ejgallego — CHANGES_REQUESTED** [review]
Code that we vendor should be marked as such and not added to our own codebase.

Embedding such code without a proper separation is bad practice and frowned upon many downstream distributions, and makes hard to backport fixes etc...

**Comment by gares** [comment]
It is 1 file belonging to the abandon ware category, IMO. Was it something real, I'd vendor it properly.

**Comment by ejgallego** [comment]
> It is 1 file belonging to the abandon ware category, IMO.

I was aware number of files determined maintenance status? What does upstream say about the maintenance status of that package?

**Comment by gares** [comment]
I'm stating two things:
- 1 file
- eject from ocaml since camlpX specific

Being it one file does make the option of depending on an external package a bit over engineered, IMO. The package also provides another file which we don't need (nor vendor in this PR).

The right fix is to avoid depending on this code. The quick one is this PR.

**Comment by SkySkimmer** [comment]
Reverted findlib bump

PS in install.md and configure we say we want ocamlfind 1.8.0 but in dockerfile we only have 1.9.1
something for another PR I guess

File: tools/coqdep/common.ml [hunk 2]
```diff
@@ -248,8 +248,10 @@ let parse_META package f =
     close_in ic;
     Some (f, m)
   with
-  | Stream.Error msg -> Error.cannot_parse_meta_file package msg
   | Sys_error _msg -> None
+  (* findlib >= 1.9.3 uses its own Error exception, so we can't catch
+     it without bumping our version requirements *)
+  | _ -> Error.cannot_parse_meta_file package
```
Comment (Alizter) [inline tools/coqdep/common.ml:254]: Can you also print the exception at least?

Comment (gares) [reply tools/coqdep/common.ml:254]: IIRC the parse error from findlib is not very different  than "error" ;-) maybe it has a line number...

Comment (Alizter) [reply tools/coqdep/common.ml:254]: Right, but this is also catching all exceptions so it could be something else.

**Comment by ejgallego** [comment]
Yes, the right fix would be to stop depending on this code; then vendoring, then this PR which increases the code we have to maintain. I think it is a bad idea to make clib / lib larger, we should aim to reduce them.

What makes depending on an external lib "over engineered"? (That part I didn't get)

**Comment by SkySkimmer** [comment]
Using a separate directory isn't any more maintenance free than putting it in clib

**Comment by Alizter** [comment]
@ejgallego Because its one file which we are planning to get rid of. If it was two or more files then it would be a different story, and vendoring would be the ideal solution.

Since the OCaml devs have basically relinquished support for this file, I agree with Enrico that this is at this point abandonware. We can maintain and eventually get rid of it if it sits in clib.

Vendoring uses the assumption that we want the package to somehow follow upstream development when needed, I don't think this is what we want. We want to get rid of it as soon as possible. As this entire PR is just a version bump, I don't think it is worth any effort setting it up as a vendored solution.

Also, as others have mentioned, we need one file from a total of 2 files in this package. Vendoring this would mean we have to keep around a useless file.

**Review by Alizter — DISMISSED** [review]

**Comment by ejgallego** [comment]
> Using a separate directory isn't any more maintenance free than putting it in clib

How so? `vendor` means not maintained by us, `clib` means maintained by clib-maintainers. As a clib maintainer I don't want even a single line more in that directory.

@Alizter , indeed if we will eventually get rid of it, how does `vendor` make a difference? If something, it serves as a reminder.

If the code is abandonware, more of a reason for me to block it arriving to Coq.

Note that setting it as a vendored solution is little effort with current build infrastructure.

**Comment by ppedrot** [comment]
I think you're overthinking it. It's a ridiculously small file that any OCaml noob can reimplement at day one. We literally have the equivalent purely functional variant IStream in clib and nobody complains. We're not pulling it an HTTP server or something, it's a trivial library.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
Using streams to parse 500kb files in 2020 on a mainstream system is likely to be slower that just reading the whole file in one go into a string and parse that.

**Comment by ejgallego** [comment]
> I think you're overthinking it. It's a ridiculously small file that any OCaml noob can reimplement at day one. We literally have the equivalent purely functional variant IStream in clib and nobody complains. We're not pulling it an HTTP server or something, it's a trivial library.

Yes I'm overthinking it and it very annoying that we do still keep insisting on doing something that is universially considered a really bad practice (cut and paste code) instead of adding a dep (like the rest of the OCaml world has done) or vendoring which adds one line of code to Coq (setting up `(vendored_dirs vendor/camlp-stream)`.

Yes we have a mess in `lib/clib` , but I fail to see how having a mess justifies adding yet more mess.

Coq situation w.r.t. external dependencies is, in my view, untenable, and should not be further promoted.

**Comment by ejgallego** [comment]
> Coq situation w.r.t. external dependencies is, in my view, untenable, and should not be further promoted

Just to remark my previous example of problematic the current approach is, take Debian. If we let this patch in, Debian packages would be forced by Debian policy to patch Coq and have it instead build againt Debian copy of camlp-streams, which is a huge load on them.

**Comment by gares** [comment]
You propose vendoring, that would require the same work on the Debian side, if the maintainer decides to do it: exceptions to the policy are tolerated, especially for non-security concerned code.

Also, coq is also a library and vendoring a library in a library is broken in Dune. I did open an issue a while ago.

Please, already 4 devs said they are ok with this pr. How many more do you need in order to let it go?

**Comment by ejgallego** [comment]
> You propose vendoring, that would require the same work on the Debian side, if the maintainer decides to do it: exceptions to the policy are tolerated, especially for non-security concerned code.

I propose in this order a) to depend on the package, b) to vendor, c) to fork and maintain ourselves.

Vendoring makes the patch to Debian much easier, but I still prefer a)

> Also, coq is also a library and vendoring a library in a library is broken in Dune. I did open an issue a while ago.

You need indeed to prefix the library, but that's easy.

> Please, already 4 devs said they are ok with this pr. How many more do you need in order to let it go?

It is not a matter of numbers, it is a matter of avoiding a very bad development practice, which as far as I have seen has not been properly supported (I can still see my questions not addressed above)

But indeed, if people want to go ahead and increase tech debt and maintainership burden, go ahead.

**Comment by ejgallego** [comment]
I'd be interested to hear any technical argument that:
 - a) supports the random code dump that `lib` / `clib` have become instead of using a proper standard library
 - b) considers increasing it is good development practice.

**Comment by Alizter** [comment]
@ejgallego Since I'm guessing everybody has an idea what vendoring is but not how to do it, can you fill us in on what needs to be done? This will be the first package vendored in Coq, so if it is as easy as you claim, it should be easy to explain no?

**Comment by SkySkimmer** [comment]
It's really a fork not a vendor

**Comment by gares** [comment]
Vendoring is in theory just composing (putting a copy of the library in your source tree). It does not work out of the box: https://github.com/ocaml/dune/issues/3335 https://github.com/ocaml/dune/issues/4070 . So let's close this branch of the discussion.

**Comment by SkySkimmer** [comment]
Why are you complaining about clib when it never causes problems?

**Comment by ejgallego** [comment]
@SkySkimmer for you it doesn't cause problems, but for me many, and it is a burden. Given that we are not converging let's discuss about this in the call.

File: clib/stream.ml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    else None)
+
+
+let of_channel ic =
+  {count = 0;
+        data = Sbuffio {ic = ic; buff = Bytes.create 4096; len = 0; ind = 0}}
+
+(* For debugging use *)
+
+let rec dump : type v. (v -> unit) -> v t -> unit = fun f s ->
```
Comment (ppedrot) [inline clib/stream.ml:161]: This is dead debug code, you should remove.

File: clib/stream.ml [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type 'a t = { mutable count : int; mutable data : 'a data }
+and 'a data =
+    Sempty
+  | Scons of 'a * 'a data
+  | Sapp of 'a data * 'a data
```
Comment (ppedrot) [inline clib/stream.ml:20]: This constructor is not used AFAICT.

File: clib/stream.ml [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  let (al, d, len) = nget n s in
+  s.count <- (s.count - len);
+  s.data <- d;
+  al
+
+let nth n st =
+  try List.nth (npeek (n+1) st) n
+  with Stdlib.Failure _ -> raise Failure
+
+let rec njunk n st =
```
Comment (ppedrot) [inline clib/stream.ml:123]: Both njunk and npeek are unused.

File: parsing/cLexer.ml [hunk 6]
```diff
@@ -487,7 +487,7 @@ let rec progress_further loc last nj tt cs =
 and update_longest_valid_token loc last nj tt cs =
   match tt.node with
   | Some _ as last' ->
-    stream_njunk nj cs;
+    Stream.njunk nj cs;
```
Comment (SkySkimmer) [inline parsing/cLexer.ml:490]: njunk

File: clib/stream.ml (hunk 5 above)
Comment (SkySkimmer) [reply clib/stream.ml:123]: npeek is used in a bunch of places (18 matches for Stream.npeek in grep)

File: clib/stream.mli [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+(** {1 Stream builders} *)
+
+val from : (unit -> 'a option) -> 'a t
+(** [Stream.from f] returns a stream built from the function [f]. To
+    create a new stream element, the function [f] is called. The user
+    function [f] must return either [Some <value>] for a value or
+    [None] to specify the end of the stream.
+*)
+
+val of_list : 'a list -> 'a t
```
Comment (ppedrot) [inline clib/stream.mli:39]: The only use of `Stream.of_list` I can find is in coqloop and is only a bad encoding of `Stream.empty`.

**Comment by ejgallego** [comment]
If I understood the discussion correctly we decided to fork and simplify the stream lib, but to make it "private" to gramlib.

@ppedrot please set the milestone you consider appropiate.

File: plugins/ltac/g_tactic.mlg [hunk 8]
```diff
@@ -59,6 +59,7 @@ open Extraargs
 
 (* idem for (x1..xn:t) [n^2 complexity but exceptional use] *)
 let check_for_coloneq =
+  let module LStream = Gramlib.LStream in
```
Comment (ejgallego) [inline plugins/ltac/g_tactic.mlg:62]: Not sure this introduces a zero-cost module alias as in toplevel `module LStream = Gramlib.LStream`

**Review by ejgallego — APPROVED** [review]
LGTM

**Comment by ejgallego** [comment]
I suggest you clean up the history a bit by amending the original commit as to move stream directly to `gramlib` (my own amend + rebase on magit makes this a painless process)

**Comment by SkySkimmer** [comment]
>I suggest you clean up the history a bit by amending the original commit as to move stream directly to gramlib (my own amend + rebase on magit makes this a painless process)

I don't know what you did but I get rebase conflicts when I try, not worth the bother IMO

**Comment by ejgallego** [comment]
> I don't know what you did but I get rebase conflicts when I try, not worth the bother IMO

I just did it, I can push if you want. It is worth IMHO as to improve blame a bit, which gets messy with the renamings.

**Comment by ejgallego** [comment]
> push on

Done, I reordered a bit while I was on it, so the PR is now logically two PRs: vendor stream + 4.14 bump

**Comment by SkySkimmer** [comment]
something weird happened with https://github.com/coq/coq/commit/b0d6e1af77f2d33766c688520b6ccd7ef4e8b74e, I fixed it

File: tools/coqdep/lib/error.ml [hunk 9]
```diff
@@ -10,7 +10,7 @@
 
 exception CannotParseFile of string * (int * int)
 exception CannotParseProjectFile of string * string
-exception CannotParseMetaFile of string * string
+exception CannotParseMetaFile of string
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply tools/coqdep/lib/error.ml:13]: changed

**Review by Alizter — APPROVED** [review]

File: dev/ci/user-overlays/15867-SkySkimmer-edge-4.14.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+overlay elpi https://github.com/SkySkimmer/coq-elpi edge-4.14 15867
+
+overlay vscoq https://github.com/SkySkimmer/vscoq edge-4.14 15867
+
+overlay gappa https://github.com/SkySkimmer/https://gitlab.inria.fr/gappa/coq edge-4.14 15867
```
Comment (Alizter) [inline dev/ci/user-overlays/15867-SkySkimmer-edge-4.14.sh:5]: Probably want to change this one

File: doc/changelog/11-infrastructure-and-dependencies/15867-edge-4.14.rst [hunk 11]
```diff
@@ -0,0 +1,4 @@
+- **Changed:**
```
Comment (Zimmi48) [inline doc/changelog/11-infrastructure-and-dependencies/15867-edge-4.14.rst:1]:
```suggestion
- **Added:**
```

**Comment by ejgallego** [comment]
We are good to go, right?

**Comment by Alizter** [comment]
@kit-ty-kate Our use of the library was small and it was decided above that it is pretty much abandonware. We therefore did not want to depend on the library externally, nor did we want to vendor 2 files needlessly. Therefore we took out the needed parts, improved others and integrated the library.

I would say this is up to coq-serapi at this point to fix this. @ejgallego What do you think?

I agree that this situation is less than ideal, but the conversion should not be very difficult.

**Comment by ejgallego** [comment]
Hi Kate, how does this code break anything in opam? So far we didn't release any of this code yet.

Coq culture about external deps is IMHO not sustainable and I have tried to point out that too, but I think that the current problem in Opam has nothing to do with this PR.

Also, I don't see how this PR creates problems in the future, as Stream is not exposed at top-level but wrapped inside the user.

**Comment by Alizter** [comment]
You can still use camlp-streams in 4.14, it will just give an alert. Past releases of Coq are able to built on 4.14 since they are building in "release mode". Could it be that the camlp-streams and stdlib parts are incompatible?

**Comment by ejgallego** [comment]
@Alizter indeed it seems to me that the bug is in camlp-stremas not on us.

**Comment by ejgallego** [comment]
Namely, Coq builds before camlp-streams is installed, then Serapi deps install it, and due to the latter overriding the global namespace (huhu 🐈‍⬛ 🐈‍⬛ ) boom! 💥 💥 💥 🐈

**Comment by Zimmi48** [comment]
>8.14

You probably mean 4.14.

**Comment by ejgallego** [comment]
Yes thanks!

**Pull request merged**